executor/predictor/metadata.go

```diff
@@ -0,0 +1,122 @@
+package predictor
+
+import (
+	"fmt"
+	"github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
+)
+
+type MetadataTensor struct {
+	DataType string `json:"datatype,omitempty"`
+	Name     string `json:"name,omitempty"`
+	Shape    []int  `json:"shape,omitempty"`
+}
+
+type ModelMetadata struct {
+	Name     string           `json:"name,omitempty"`
+	Platform string           `json:"platform,omitempty"`
+	Versions []string         `json:"versions,omitempty"`
+	Inputs   []MetadataTensor `json:"inputs,omitempty"`
+	Outputs  []MetadataTensor `json:"outputs,omitempty"`
+}
+
+type GraphMetadata struct {
+	Name         string                   `json:"name,omitempty"`
+	Models       map[string]ModelMetadata `json:"models,omitempty"`
+	GraphInputs  []MetadataTensor         `json:"graphinputs,omitempty"`
+	GraphOutputs []MetadataTensor         `json:"graphoutputs,omitempty"`
+}
+
+func NewGraphMetadata(p *PredictorProcess, spec *v1.PredictorSpec) (output *GraphMetadata, err error) {
+	output = &GraphMetadata{}
+	output.Models, err = p.MetadataMap(spec.Graph)
+	if err != nil {
+		return nil, err
+	}
+	output.Name = spec.Name
+	output.GraphInputs, output.GraphOutputs = output.GetShapeFromGraph(spec.Graph)
+	return
+}
+
+func (p *GraphMetadata) GetShapeFromGraph(node *v1.PredictiveUnit) (
+	input []MetadataTensor, output []MetadataTensor,
+) {
+	nodeMeta := p.Models[node.Name]
+	nodeInputs := nodeMeta.Inputs
+	nodeOutputs := nodeMeta.Outputs
+
+	// Node has no children
+	if node.Children == nil || len(node.Children) == 0 {
+		// If Node is model then inputs and outputs are clear
+		if *node.Type == v1.MODEL {
+			return nodeInputs, nodeOutputs
+		} else {
+			fmt.Println("Unkown case.")
+			return nil, nil
```

Comment:
There are quite a few places in this function where it silently returns `nil` for both return parameters. Could this be considered an error in any of these cases?

Reply:
Could be but first we should define a desired behaviour for corner cases.

There could be situations when for example three-level graph (1->2->3) have metadata missing for `2` or output of `1` does not match input of `2`. We could anyway define global input (1) and output (3) here... but some warnings should be raised because either metadata is corrupted or wrong models were used in the graphs.

Reply:
With @axsaucedo we decided to skip testing for corner cases currently.